cpu/gd32v/periph/gpio.c

```diff
@@ -0,0 +1,243 @@
+ */
+#define MODE_MASK                       (0x0f)
+#define ODR_POS                         (4U)
+
+/**
+ * @brief   Extract the port base address from the given pin identifier
+ */
+static inline GPIO_Type *_port(gpio_t pin)
+{
+    return (GPIO_Type*)(pin & ~(0x0f));
+}
+
+/**
+ * @brief   Extract the port number form the given identifier
+ *
+ * The port number is extracted by looking at bits 10, 11, 12, 13 of the base
+ * register addresses.
+ */
+static inline int _port_num(gpio_t pin)
+{
+    return (((pin >> 10) & 0x0f) - 2);
+}
+
+/**
+ * @brief   Extract the pin number from the last 4 bit of the pin identifier
+ */
+static inline int _pin_num(gpio_t pin)
+{
+    return (pin & 0x0f);
+}
+
+static inline void _port_enable_clock(gpio_t pin)
+{
+    periph_clk_en(APB2, (RCU_APB2EN_PAEN_Msk << _port_num(pin)));
+}
+
+/**
+ * @brief   Check if the given mode is some kind of input mdoe
+ * @param[in]   mode    Mode to check
+ * @retval  1           @p mode is GPIO_IN, GPIO_IN_PD, or GPIO_IN_PU
+ * @retval  0           @p mode is not an input mode
+ */
+static inline int gpio_mode_is_input(gpio_mode_t mode)
+{
+    return !(mode & 3);
+}
+
+static inline void _set_mode_or_af(GPIO_Type *port, unsigned pin_num,
+                                   unsigned mode_or_af)
+{
+    volatile uint32_t *crl = (&port->CTL0 + (pin_num >> 3));
+    uint32_t tmp = *crl;
+    tmp &= ~(0xf << ((pin_num & 0x7) * 4));
+    tmp |= ((mode_or_af & MODE_MASK) << ((pin_num & 0x7) * 4));
+    *crl = tmp;
+}
+
+static inline bool _pin_is_output(GPIO_Type *port, unsigned pin_num)
+{
+    uint32_t reg = *(uint32_t *)(&port->CTL0 + (pin_num >> 3));
+    return reg & (0x3 << ((pin_num & 0x7) << 2));
+}
+
+int gpio_init(gpio_t pin, gpio_mode_t mode)
+{
+    GPIO_Type *port = _port(pin);
+    unsigned pin_num = _pin_num(pin);
+
+    /* open-drain output with pull-up is not supported */
+    if (mode == GPIO_OD_PU) {
+        return -1;
+    }
+
+    /* enable the clock for the selected port */
+    _port_enable_clock(pin);
+
+    /* set pin mode */
+    _set_mode_or_af(port, pin_num, mode);
+
+    /* For input modes, ODR controls pull up settings */
+    if (gpio_mode_is_input(mode)) {
+        if (mode == GPIO_IN_PU) {
+            cpu_reg_enable_bits(&port->OCTL, 1 << pin_num);
+        }
+        else {
+            cpu_reg_disable_bits(&port->OCTL, 1 << pin_num);
+        }
+    }
+
+    return 0; /* all OK */
+}
+
+void gpio_init_af(gpio_t pin, gpio_af_t af)
+{
+    GPIO_Type *port = _port(pin);
+    unsigned pin_num = _pin_num(pin);
+
+    /* enable the clock for the selected port */
+    _port_enable_clock(pin);
+    /* configure the pin */
+    _set_mode_or_af(port, pin_num, af);
+}
+
+void gpio_init_analog(gpio_t pin)
+{
+    /* enable the GPIO port RCC */
+    _port_enable_clock(pin);
+
+    /* map the pin as analog input */
+    int pin_num = _pin_num(pin);
+    *(uint32_t *)(&_port(pin)->CTL0 + (pin_num >= 8)) &= ~(0xfl << (4 * (pin_num - ((pin_num >= 8) * 8))));
```

Comment:
```suggestion
    volatile uint32_t *foo = (uint32_t)(&_port(pin)->CTL0 + (pin_num >= 8));
    foo &= ~(0xfl << (4 * (pin_num - ((pin_num >= 8) * 8))));
```

This would be more readable when split into two statements. I also think that you need a `volatile` keyword here, as this write access likely will have side affects that the C compiler will not know about. (And a better name than `foo` might also provide enough info for the causal reader to understand this. If not, maybe add a comment?)